pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go

```diff
@@ -242,6 +229,28 @@ func (c *Controller) patchCondition(ctx context.Context, old, new *tenancyv1alph
 	return err
 }
 
+func (c *Controller) patchFinalizer(ctx context.Context, cluster logicalcluster.Name, name string, finalizers []string) error {
+	finalizerData := &metav1.PartialObjectMetadata{
+		ObjectMeta: metav1.ObjectMeta{
+			Finalizers: finalizers,
+		},
+	}
+
+	patch, err := json.Marshal(finalizerData)
+	if err != nil {
+		return err
+	}
+
+	// remove finalizers field if there is no remaining finalizers,
+	if len(finalizers) == 0 {
+		patch = []byte("{\"metadata\": {\"finalizers\": []}}")
```

Comment:
not sure this is safe. Removing the field is not really necessary. Just merge-patch your finalizer out. This patch here has the potential to override other finalizers just added. 

same for the same func above.

Reply:
added uid/rv as precond in this case. The finalizer is opmitempty, so I think we have to build this json string explicitly when finalizer is empty.